src/backends/mock_main.rs

```diff
@@ -35,18 +37,103 @@ impl OperationArg {
 struct Operation(pub NativeOperation, pub Vec<OperationArg>);
 
 impl Operation {
-    pub fn deref(&self, statements: &[Statement]) -> crate::middleware::Operation {
+    pub fn deref(&self, statements: &[Statement]) -> Result<crate::middleware::Operation> {
         let deref_args = self
             .1
             .iter()
-            .map(|arg| match arg {
-                OperationArg::None => middleware::OperationArg::None,
+            .flat_map(|arg| match arg {
+                OperationArg::None => vec![],
                 OperationArg::Index(i) => {
-                    middleware::OperationArg::Statement(statements[*i].clone())
+                    vec![statements[*i].clone().try_into()]
                 }
             })
-            .collect();
-        middleware::Operation(self.0, deref_args)
+            .collect::<Result<Vec<crate::middleware::Statement>>>()?;
+        middleware::Operation::op(self.0, &deref_args)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Statement(pub NativeStatement, pub Vec<StatementArg>);
+
+impl Statement {
+    pub fn is_none(&self) -> bool {
+        self.0 == NativeStatement::None
+    }
+}
+
+impl ToFields for Statement {
+    fn to_fields(self) -> (Vec<middleware::F>, usize) {
+        let (native_statement_f, native_statement_f_len) = self.0.to_fields();
+        let (vec_statementarg_f, vec_statementarg_f_len) = self
+            .1
+            .into_iter()
+            .map(|statement_arg| statement_arg.to_fields())
+            .fold((Vec::new(), 0), |mut acc, (f, l)| {
+                acc.0.extend(f);
+                acc.1 += l;
+                acc
+            });
+        (
+            [native_statement_f, vec_statementarg_f].concat(),
+            native_statement_f_len + vec_statementarg_f_len,
+        )
+    }
+}
+
+impl TryInto<middleware::Statement> for Statement {
+    type Error = anyhow::Error;
+    fn try_into(self) -> Result<middleware::Statement> {
+        type S = middleware::Statement;
+        type NS = NativeStatement;
+        type SA = StatementArg;
+        let args = (
+            self.1.get(0).cloned(),
+            self.1.get(1).cloned(),
+            self.1.get(2).cloned(),
+        );
```

Comment:
Maybe we also need to include `args.len()` in the match to validate the arguments length of each statement